### DISCUSSION TITLE
Getting a nice-flowing docker dev setup

### DISCUSSION THREAD

**Discussion by LavissaWoW** [inventree/InvenTree#4731] [Q&A]
I'm having some internittent struggles getting a proper flowing dev setup working...
My main dev setup involves Windows, as that's what my home PC has. Linux will be intermittently used when I have time during work hours. (I don't suspect I'll have issues there)

Notes:
- Windows
- VS Code
- Docker
- Extra WSL terminal for grep etc

Now, I've set up such that I clone from my fork down to my machine using Windows CMD, as that's where I have my ssh key.
Then I set up InvenTree dev on Docker via the docker compose [...] update / setup-dev / up -d / superpuser commands in that order (In the Windows CMD).

Now, we're in a bit of a pickle. I don't want to create SSH keys for docker containers that might be terminated at any time. So I can't set up the docker to do commit/push/pull in any easy fashion. 
Venv is accessible via WSL (as the  docker image does not set up a Windows activate.bat), but triggers permission denied when run via the docker terminal. Venv  in WSL is however not privvy to installed packages on the docker, so can't execute invokes.

I attempted a setup using the supplied config for VSCode devcontainer, and while that works 80%, it still won't cooperate with git. Additionally, the VSCode terminal sometimes freezes in devcontainer, and I'm forced to restart VSCode. Additionally, additionally, I dislike having to wait for the container to boot, load etc every time I open VSCode.

Atm, my workflow is:

- VSCode with many, many warnings of nonexistent django packages
- VSCode does have git tracking in this mode, which I like to have
- To use pre-commit, I have to issue commits in WSL with pre-commit installed
- Then re-issue said commit in Windows with --no-verify due to ssh keys. 

Any tips or tricks to achieve a more coherent workflow?

**Comment by SchrodingersGat** [comment]
@LavissaWoW have you followed the [docker development guide](https://docs.inventree.org/en/latest/start/docker_dev/)? This is exactly how my development setup is configured, I'm running on Windows too.

I edit in vscode but have the containers managed through docker desktop. With the vscode terminal I can oversee the containers (e.g. `docker compose run inventree-dev-server invoke migrate`)

**Reply by wolflu05** [reply]
But this won't add python intelligence about external modules and if you don't have python installed locally you don't have any intelligence. Devcontainer is pretty good about that, only downside for me (which I didn't managed to fix where I initially implemented it) that gpg keys and git config are not available inside the container, so I need to commit outside, but there the pre commit hooks don't work.

**Reply by SchrodingersGat** [reply]
@wolflu05 good point. We should definitely improve our vscode / devcontainer docs, as this is pretty much standard approach now (although I myself haven't fully moved across to using it yet).

we have an open issue here to improve the docs, if anyone wants to step up to the plate, it would be greatly appreciated:

- https://github.com/inventree/InvenTree/issues/4670
- https://github.com/inventree/inventree-docs/issues/350
- https://github.com/inventree/InvenTree/pull/3474

**Reply by LavissaWoW** [reply]
I've followed it, yes. 
Just to be clear, it mostly works as one would want. It's the venv and git issue that's clogging up the flow. (And VSCode's inability to detect external modules generates IDE noise)
pre-commit does not exist by default in Windows, so a dev setup following the docs will not allow you to perform commits atm. 
You won't be able to commit inside a WSL venv if you pulled in Windows. Basically, it's the same issue as with the devcontainer that @wolflu05 mentions. Can't commit in-scope due to git config and keys, and can't commit out-of-scope due to inability to run pre-commit hooks.


I've also made a couple of observations as to setting it up that I think would be decent to alter.

- Every `docker compose run inventree-dev-server [command]` generates a new inventree-dev-server container which never gets used again after executing its command and exiting (At least locally on my computer). As a result, the list of containers under inventree in Docker Desktop grows for each run. `docker compose run --rm [...]` will alleivate this, as the command container will be deleted upon completion.
- setup-test does not by default invoke setup-dev. This means you're going to be missing things from requirements-dev. Amending `docker compose run inventree-dev-server invoke setup-test` with --dev will solve this issue.
- I've had some odd issues with superuser, where the user isn't created. Atm, it looks like superuser must be run after `docker compose up -d`. I'll investigate that a little bit further though.
I can make some changes in the docs and put up a PR to review changes

**Reply by wolflu05** [reply]
I think there is a much easier way to get gpg keys and git config working in a devcontainer with vscode than getting it working in a normally used container, but I didn't spend that much time to come up with something working generically and not hacky.


> Every docker compose run inventree-dev-server [command] generates a new inventree-dev-server container which never gets used again after executing its command and exiting (At least locally on my computer). As a result, the list of containers under inventree in Docker Desktop grows for each run. docker compose run --rm [...] will alleivate this, as the command container will be deleted upon completion.

You can use docker-compose exec to get a shell into an already runny container.

**Reply by LavissaWoW** [reply]
@wolflu05 I absolutely presume there's a way to use exec instead of run, though I'm not gonna research it rn.
I quickly tried running `docker compose exec inventree-dev-server invoke test` in place of `docker compose run inventree-dev-server invoke test`, and that immediately went into "module django not found" where the latter command executes what you want, but ofc creates a new throwaway container.

For the install process, we don't have any running containers yet, so it makes sense to use `run --rm` to purge these containers imo.

**Reply by SchrodingersGat** [reply]
Good point about `--rm` - we should update the docs for *some* of the commands

**Reply by wolflu05** [reply]
@SchrodingersGat See https://github.com/inventree/InvenTree/pull/4743#discussion_r1181721690 for a discussion why we shouldn't do that.

**Reply by SchrodingersGat** [reply]
@wolflu05 yep thanks I found that comment thread after this one :)

**Comment by SchrodingersGat** [comment]
@LavissaWoW please keep us posted if you come up with any improvements to the workflow

**Comment by wolflu05** [comment]
@LavissaWoW May you can take a look at #4787 - I tried to document the devcontainer workflow there.

**Reply by LavissaWoW** [reply]
I'll give it a whirl
